geometric_kernels/spaces/so.py

```diff
@@ -0,0 +1,368 @@
+            chi_variables = gammas + gammas_conj
+            chi_poly = sympy.Poly(chi_poly.subs([gr ** 2, g] for gr, g in zip(gammas_sqrt + gammas_conj_sqrt, chi_variables)), chi_variables)
+        p = sympy.Poly(chi_poly.subs((c, c_val) for c, c_val in zip(chi_coeffs, sol)), chi_variables)
+        coeffs = list(map(int, p.coeffs()))
+        monoms = [list(map(int, monom)) for monom in p.monoms()]
+        return coeffs, monoms
+
+    def _inverse(self, X: B.Numeric) -> B.Numeric:
+        return B.inverse(X)
+
+    def num_levels(self) -> int:
+            """Number of levels, L"""
+            return self._num_levels
+
+    def num_eigenfunctions_per_level(self) -> int:
+        """Number of eigenfunctions per level
+        Is it right?"""
+        return self._dimensions
+
+
+class SOCharacter(LieGroupCharacter):
+    def __init__(self, n, signature):
+        self.signature = signature
+        self.n = n
+        self.coeffs, self.monoms = self._load()
+
+    def _load(self):
+        group_name = '{SO}({})'.format(self.n)
+        file_path = Path(__file__).with_name('precomputed_characters.json')
+        with file_path.open('r') as file:
+            character_formulas = json.load(file)
+            try:
+                cs, ms = character_formulas[group_name][str(self.signature)]
+                self.coeffs, self.monoms = (np.array(data) for data in (cs, ms))
+            except KeyError as e:
+                raise KeyError('Unable to retrieve character parameters for signature {} of {}, '
+                               'perhaps it is not precomputed.'.format(e.args[0], group_name)) from None
+
+    def __call__(self, gammas):
+        gammas = B.cat((gammas, gammas.conj()), dim=-1)
+        char_val = B.zeros(B.dtype(gammas), gammas.shape[:-1])
+        for coeff, monom in zip(self.coeffs, self.monoms):
+            char_val += coeff * B.prod(gammas ** from_numpy(gammas, monom), dim=-1)
+        return char_val
+
+
+class SOGroup(LieGroup, gs.geometry.special_orthogonal):
+    r"""
+    The d-dimensional hypersphere embedded in the (d+1)-dimensional Euclidean space.
+    """
+    def __init__(self, n):
+        self.n = n
+        self.dim = n * (n-1) // 2
+
+    def random(self, key, number):
+        if self.n == 2:
+            # SO(2) = S^1
+            thetas = 2 * math.pi * B.random.randn(key, dtype_double(key), number, 1)
+            c = B.cos(thetas)
+            s = B.sin(thetas)
+            r1 = B.hstack((c, s)).unsqueeze(-2)
+            r2 = B.hstack((-s, c)).unsqueeze(-2)
+            q = B.cat((r1, r2), dim=-2)
+            return q
+        elif self.n == 3:
+            # explicit parametrization via the double cover SU(2) = S^3
+            sphere_point = B.random.randn(key, dtype_double(key), number, 4)
+            sphere_point /= B.linalg.vector_norm(sphere_point, dim=-1, keepdim=True)
+            x, y, z, w = (sphere_point[..., i].unsqueeze(-1) for i in range(4))
+            xx = x ** 2
```

Comment:
this is a minor thing but: maybe compress these lines into three with the commas? 
To my eye, this just looks like a wall